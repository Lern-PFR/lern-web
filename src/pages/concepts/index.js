import ConceptDetailsPage from './ConceptDetailsPage';

export {
	// eslint-disable-next-line import/prefer-default-export
	ConceptDetailsPage,
};
