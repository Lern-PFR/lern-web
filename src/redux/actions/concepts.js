import * as ConceptsApi from 'api/conceptsApi';
import i18next from 'i18next';
import { toast } from 'react-toastify';
import { fetchSubject } from './subjects';

/**
 * @constant
 * @name ActionTypes
 * @description The various action types used to interact with the concepts redux state.
 * @type {object}
 *
 * @author Timothée Simon-Franza
 */
export const ActionTypes = {
	FETCH_CONCEPT_REQUEST: '@CONCEPTS/FETCH_REQUEST',
	FETCH_CONCEPT_SUCCESS: '@CONCEPTS/FETCH_SUCCESS',
	FETCH_CONCEPT_FAILURE: '@CONCEPTS/FETCH_FAILURE',

	FETCH_CONCEPT_LIST_REQUEST: '@CONCEPTS/FETCH_LIST_REQUEST',
	FETCH_CONCEPT_LIST_SUCCESS: '@CONCEPTS/FETCH_LIST_SUCCESS',
	FETCH_CONCEPT_LIST_FAILURE: '@CONCEPTS/FETCH_LIST_FAILURE',

	CREATE_CONCEPT_REQUEST: '@CONCEPTS/CREATE_REQUEST',
	CREATE_CONCEPT_SUCCESS: '@CONCEPTS/CREATE_SUCCESS',
	CREATE_CONCEPT_FAILURE: '@CONCEPTS/CREATE_FAILURE',

	UPDATE_CONCEPT_REQUEST: '@CONCEPTS/UPDATE_REQUEST',
	UPDATE_CONCEPT_SUCCESS: '@CONCEPTS/UPDATE_SUCCESS',
	UPDATE_CONCEPT_FAILURE: '@CONCEPTS/UPDATE_FAILURE',

	DELETE_CONCEPT_REQUEST: '@CONCEPTS/DELETE_REQUEST',
	DELETE_CONCEPT_SUCCESS: '@CONCEPTS/DELETE_SUCCESS',
	DELETE_CONCEPT_FAILURE: '@CONCEPTS/DELETE_FAILURE',

	CLEAR_CONCEPT_LIST: '@CONCEPTS/CLEAR_LIST',
};

// //////////////////////////////////////////////////////// //
// //////////// Specific concept fetching actions ///////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchConceptRequest
 * @description Action triggered anytime a concept fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchConceptRequest = () => ({ type: ActionTypes.FETCH_CONCEPT_REQUEST });

/**
 * @function
 * @name fetchConceptSuccess
 * @description Action triggered as a result to a successful concept fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} concept The concept retrieved from the API.
 *
 * @returns {object}
 */
const fetchConceptSuccess = ({ concept }) => ({
	type: ActionTypes.FETCH_CONCEPT_SUCCESS,
	payload: { concept },
});

/**
 * @function
 * @name fetchConceptFailure
 * @description Action triggered as a result to a failed concept fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchConceptFailure = (error) => ({
	type: ActionTypes.FETCH_CONCEPT_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// ///////////// Concept list fetching actions //////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchConceptListRequest
 * @description Action triggered anytime a concept list fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchConceptListRequest = () => ({ type: ActionTypes.FETCH_CONCEPT_LIST_REQUEST });

/**
 * @function
 * @name fetchConceptListSuccess
 * @description Action triggered as a result to a successful concept list fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {array}	concepts	The list of concepts retrieved from the API.
 * @param {number}	totalCount	The total amount of concepts available for the module.
 *
 * @returns {object}
 */
const fetchConceptListSuccess = ({ concepts, totalCount }) => ({
	type: ActionTypes.FETCH_CONCEPT_LIST_SUCCESS,
	payload: { concepts, totalCount },
});

/**
 * @function
 * @name fetchConceptListFailure
 * @description Action triggered as a result to a failed concept list fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchConceptListFailure = (error) => ({
	type: ActionTypes.FETCH_CONCEPT_LIST_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// /////////////// Concept creation actions /////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name createConceptRequest
 * @description Action triggered anytime a concept creation call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const createConceptRequest = () => ({ type: ActionTypes.CREATE_CONCEPT_REQUEST });

/**
 * @function
 * @name createConceptSuccess
 * @description Action triggered as a result to a successfult concept creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} concept The newly created concept.
 *
 * @returns {object}
 */
const createConceptSuccess = ({ concept }) => ({
	type: ActionTypes.CREATE_CONCEPT_SUCCESS,
	payload: { concept },
});

/**
 * @function
 * @name createConceptFailure
 * @description Action triggered as a result to a failed concept creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error The exception sent back from the API.
 *
 * @returns {object}
 */
const createConceptFailure = (error) => ({
	type: ActionTypes.CREATE_CONCEPT_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// /////////////// Concept edition actions //////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name updateConceptRequest
 * @description Action triggered anytime a concept edition call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const updateConceptRequest = () => ({ type: ActionTypes.UPDATE_CONCEPT_REQUEST });

/**
 * @function
 * @name updateConceptSuccess
 * @description Action triggered as a result to a successful concept edition API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} concept The updated concept object.
 *
 * @returns {object}
 */
const updateConceptSuccess = ({ concept }) => ({
	type: ActionTypes.UPDATE_CONCEPT_SUCCESS,
	payload: { concept },
});

/**
 * @function
 * @name updateConceptFailure
 * @description Action triggered as a result to a failed concept edition API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error The exception sent back from the API.
 *
 * @returns {object}
 */
const updateConceptFailure = (error) => ({
	type: ActionTypes.UPDATE_CONCEPT_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// /////////////// Concept deletion actions /////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name deleteConceptRequest
 * @description Action triggered anytime a concept deletion call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const deleteConceptRequest = () => ({ type: ActionTypes.DELETE_CONCEPT_REQUEST });

/**
 * @function
 * @name deleteConceptSuccess
 * @description Action triggered as a result to a successful concept deletion API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} concept The deleted concept object.
 *
 * @returns {object}
 */
const deleteConceptSuccess = ({ concept }) => ({
	type: ActionTypes.DELETE_CONCEPT_SUCCESS,
	payload: { concept },
});

/**
 * @function
 * @name deleteConceptFailure
 * @description Action triggered as a result to a failed concept deletion API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const deleteConceptFailure = (error) => ({
	type: ActionTypes.DELETE_CONCEPT_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// ///////////// Concept list clearing actions //////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name clearConceptListAction
 * @description Action triggered anytime the concept list is not needed anymore.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const clearConceptListAction = () => ({ type: ActionTypes.CLEAR_CONCEPT_LIST });

// //////////////////////////////////////////////////////// //
// //////////////// Exported action creators ////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchConcept
 * @description Method used to fetch the latest version of a specified concept.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} conceptId The id of the concept we want to retrieve.
 */
export const fetchConcept = (conceptId) => (dispatch) => {
	dispatch(fetchConceptRequest());

	return ConceptsApi.fetchConceptById(conceptId)
		.then((concept) => dispatch(fetchConceptSuccess({ concept })))
		.catch((error) => dispatch(fetchConceptFailure(error)));
};

/**
 * @function
 * @name fetchConceptListByModuleId
 * @description Method used to fetch all concepts from a specific module, identified by the moduleId param.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} moduleId The id of the module we want to retrieve concepts from.
 */
export const fetchConceptListByModuleId = (moduleId) => (dispatch) => {
	dispatch(fetchConceptListRequest());

	return ConceptsApi.fetchConceptsByModuleId(moduleId)
		.then(({ concepts, totalCount }) => dispatch(fetchConceptListSuccess({ concepts, totalCount })))
		.catch((error) => dispatch(fetchConceptListFailure(error)));
};

/**
 * @function
 * @name createConcept
 * @description Method used to create a new concept in the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} conceptData The data to create the new concept from.
 */
export const createConcept = (conceptData) => (dispatch) => {
	dispatch(createConceptRequest());

	return ConceptsApi.createConcept(conceptData)
		.then(({ concept }) => dispatch(createConceptSuccess({ concept })))
		.catch((error) => dispatch(createConceptFailure(error)));
};

/**
 * @function
 * @name updateConcept
 * @description Method used to update an existing concept instance from the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} conceptData	The data to update the concept with.
 * @param {string} conceptId	The id to identify the concept to update.
 */
export const updateConcept = (conceptData, conceptId) => (dispatch) => {
	dispatch(updateConceptRequest());

	return ConceptsApi.updateConcept(conceptData, conceptId)
		.then(({ concept }) => dispatch(updateConceptSuccess({ concept })))
		.catch((error) => dispatch(updateConceptFailure(error)));
};

/**
 * @function
 * @name deleteConcept
 * @description Method used to remove an existing concept from the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} conceptId	The id identifying the concept to remove.
 * @param {string} subjectId	The id identifying the concept's parent subject.
 */
export const deleteConcept = (conceptId, subjectId) => (dispatch) => {
	dispatch(deleteConceptRequest());

	return ConceptsApi.deleteConcept(conceptId)
		.then((concept) => {
			dispatch(deleteConceptSuccess({ concept }));
			toast.success(i18next.t('concepts.deletion.toasts.success', { name: concept.title }));
			dispatch(fetchSubject(subjectId));
		})
		.catch((error) => dispatch(deleteConceptFailure(error)));
};

/**
 * @function
 * @name clearConceptList
 * @description Method used to clear the concept list stored in memory when the application does not need it anymore.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {Promise.resolve}
 */
export const clearConceptList = () => (dispatch) => {
	dispatch(clearConceptListAction());

	return Promise.resolve();
};
