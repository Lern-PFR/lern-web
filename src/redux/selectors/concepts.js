import { sortBy } from 'lodash';
import { createSelector } from 'reselect';

/**
 * @function
 * @name getConcepts
 * @description A selector callback which returns all concepts from the Redux state.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} state The redux state at the time of call.
 *
 * @returns {Array}
 */
const getConcepts = (state) => state?.concepts?.items ?? [];

/**
 * @function
 * @name getConceptById
 * @description A selector callback which returns the concept whose id matches the conceptId parameter.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} conceptId The id of the concept we want to retrieve.
 *
 * @returns {object|undefined}
 */
const getConceptById = createSelector(
	getConcepts,
	(_, conceptId) => conceptId,
	(concepts, conceptId) => {
		const concept = concepts.filter(({ id }) => (id === conceptId))?.[0] ?? undefined;

		if (concept === undefined) {
			return concept;
		}

		// @TODO: replace concept.courses with courses.lessons once the API has been updated.
		// Removes older versions of each lesson from the list.
		const lessons = Object.values((concept.courses || []).reduce((acc, currentLesson) => {
			const currentlyStoredLesson = acc[currentLesson.id];
			if (currentlyStoredLesson) {
				acc[currentLesson.id] = currentlyStoredLesson.version < currentLesson.version ? currentLesson : currentlyStoredLesson;
			} else {
				acc[currentLesson.id] = currentLesson;
			}

			return acc;
		}, {}));

		// @TODO: remove following snippet with the commented "return" snippet once the API has been updated.
		const { courses, ...result } = concept;

		return {
			...result,
			lessons: sortBy(lessons, 'order'),
		};

		// return {
		// 	...concept,
		//	lessons: sortBy(lessons),
		// };
	}
);

export {
	getConcepts,
	getConceptById,
};
