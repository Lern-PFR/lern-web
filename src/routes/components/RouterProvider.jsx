import PropTypes from 'prop-types';
import { Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';

/**
 * @name history
 * @description The global history object, handling the navigation for the whole application
 *
 * @author Yann Hodiesne
 */
export const history = createBrowserHistory();

/**
 * @name RouterProvider
 * @description A component providing a Router from react-router, with the ability to export its history and block navigation
 *
 * @author Yann Hodiesne
 *
 * @param {*} children : The children to render inside the Router
 */
const RouterProvider = ({ children }) => (
	<Router history={history}>
		{children}
	</Router>
);

RouterProvider.propTypes = {
	children: PropTypes.node.isRequired,
};

export default RouterProvider;
