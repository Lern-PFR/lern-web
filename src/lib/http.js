import localstorage from './localstorage';
import history from './history'
import { logoutAction } from '../redux/actions/auth';
import store from '../redux/store/configureStore';

const { dispatch, getState } = store;

/**
 * Generates HTTP headers.
 * 
 * If the user is authenticated, retrieves its jwt authentication token and add it to the header list.
 */
const getHeaders = () => {
    const headers = new Headers();
    const user = JSON.parse(localstorage.getItem('user'));
    
    headers.append('Content-Type', 'application/json');
  
    if (user && user.token) {
        headers.append('Authorization', `Bearer ${user.token}`);
    }
  
    return headers;
};

/**
 * Returns a query string from the object in parameter.
 * 
 * @param {object} params : The object to create the query string from.
 */
const objectToQS = (params) => Object
    .entries(params)
    .map(([key, value]) => {
        if (value !== undefined && value !== null) {
            return `${encodeURIComponent(key)}=${encodeURIComponent(value)}`;
        }

        return `${encodeURIComponent(key)}`;
    })
    .join('&');

/**
 * Generates a POST HTTP query from the url and params object in parameters.
 * 
 * @param {string} url      : The url to target.
 * @param {object} params   : The parameters to inject into the query.
 */
export const post = (url, params) => fetch(url, {
    method: 'POST',
    body: JSON.stringify(params),
    headers: getHeaders(),
});

/**
 * Generates a GET HTTP query from the url and params object in parameters.
 * 
 * If the params object is set, converts its content into a query string.
 * 
 * @param {string} url      : The url to target.
 * @param {object} params   : The parameters to inject into the query.
 */
export const get = (url, params) => {
    const qs = params ? `?${objectToQS(params)}` : '';

    return fetch(`${url}${qs}`, {
        headers: getHeaders(),
    });
};

/**
 * Generates a PUT HTTP query from the url and params object in parameters.
 * 
 * @param {string} url       : The url to target.
 * @param {object} params    : The parameters to inject into the query.
 */
export const put = (url, params) => fetch(url, {
    method: 'PUT',
    body: JSON.stringify(params),
    headers: getHeaders(),
});

/**
 * Generates a DELETE HTTP query from the url in parameters.
 * 
 * @param {string} url 
 */
export const del = (url) => fetch(url, {
    method: 'PUT',
    headers: getHeaders(),
});

/**
 * Handles the response from an HTTP query.
 * 
 * If a trigger is passed to the method, dispatch it using redux.
 * 
 * @param {*} trigger : The redux action to dispatch.
 */
export const handleResponse = (trigger) => (response) => {
    if (response.status === 204) {
      if (trigger) {
        dispatch(trigger());
      }
  
      return Promise.resolve();
    }
  
    if (response.status === 403 || response.status === 413 || response.status >= 500) {
      return Promise.reject(response);
    }
  
    return response.json().then((res) => {
        if (response.status === 401 && res.message !== 'Bad credentials') {
            dispatch(logoutAction());
            history.push('/');
    
            return Promise.reject({
                ...res,
                status: response.status,
            });
        }
    
        if (response.status >= 300) {
            return Promise.reject({
                ...res,
                status: response.status,
            });
        }
    
        if (trigger) {
            dispatch(trigger(res));
        }
    
        return Promise.resolve(res);
    });
};