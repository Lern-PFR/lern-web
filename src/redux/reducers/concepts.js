import { ActionTypes } from 'redux/actions/concepts';

const initialState = {
	isLoading: false,
	items: [],
	totalCount: 0,
};

export default (state = initialState, { type, payload }) => {
	switch (type) {
		case ActionTypes.FETCH_CONCEPT_REQUEST:
		case ActionTypes.FETCH_CONCEPT_LIST_REQUEST:
		case ActionTypes.CREATE_CONCEPT_REQUEST:
		case ActionTypes.UPDATE_CONCEPT_REQUEST:
		case ActionTypes.DELETE_CONCEPT_REQUEST:
			return {
				...state,
				isLoading: true,
			};
		case ActionTypes.FETCH_CONCEPT_FAILURE:
		case ActionTypes.FETCH_CONCEPT_LIST_FAILURE:
		case ActionTypes.CREATE_CONCEPT_FAILURE:
		case ActionTypes.UPDATE_CONCEPT_FAILURE:
		case ActionTypes.DELETE_CONCEPT_FAILURE:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.CREATE_CONCEPT_SUCCESS:
		case ActionTypes.DELETE_CONCEPT_SUCCESS:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.FETCH_CONCEPT_SUCCESS:
		case ActionTypes.UPDATE_CONCEPT_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: [payload.concept],
				totalCount: 1,
			};
		case ActionTypes.FETCH_CONCEPT_LIST_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: [...payload.concepts],
				totalCount: payload.totalCount,
			};
		case ActionTypes.CLEAR_CONCEPT_LIST:
			return {
				...state,
				items: [],
				totalCount: 0,
			};
		default:
			return state;
	}
};
