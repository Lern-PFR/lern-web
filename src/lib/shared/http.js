import { history } from 'routes/components/RouterProvider';
import session from 'lib/shared/session';
import routes from 'routes';

/**
 * @constant
 * @name baseUrl
 * @description The base url to append to in order to target and API endpoint.
 * @type {string}
 *
 * @author Timothée Simon-Franza
 *
 * Note : Its value is set using environment variables set defined in this .env file and forwarded by docker-compose.
 */
export const baseUrl = process.env.REACT_APP_API_URL;

/**
 * @function
 * @name getHeaders
 * @description Generates and returns a header bundle from the current session's data.
 *
 * @author Timothée Simon-Franza
 * @author Yann Hodiesne
 *
 * @returns {Header} the generated header bundle.
 */
export const getHeaders = () => {
	const headers = new Headers();

	headers.append('Content-Type', 'application/json');

	if (session.exists()) {
		headers.append('Authorization', `Bearer ${session.get()}`);
	}

	return headers;
};

/**
 * @function
 * @name objectToQS
 * @description Converts the object in parameter to a query string.
 *
 * @author Timothée Simon-Franza
 * @author Yann Hodiesne
 *
 * @param {object} params : The object to convert to QS.
 *
 * @returns {string}
 */
export const objectToQS = (params) => Object
	.entries(params)
	.map(([key, value]) => {
		if (value !== undefined && value !== null) {
			return `${encodeURIComponent(key)}=${encodeURIComponent(value)}`;
		}

		return `${encodeURIComponent(key)}`;
	})
	.join('&');

/**
 * @function
 * @name handleResponse
 * @description Automatically deserializes and handles the received response object.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} response : The response to handle.
 *
 * @returns {Promise | object}
 */
export const handleResponse = (response) => {
	if (response.status === 204) {
		return Promise.resolve();
	}

	if (response.status === 401 && session.exists()) {
		history.push(routes.auth.logout);
	}

	if (response.status === 401 || response.status === 403 || response.status === 409 || response.status === 413 || response.status >= 500) {
		return Promise.reject(response);
	}

	return response.json();
};

/**
 * @function
 * @name handleError
 * @description Automatically handles the received error to return a formatted object.
 *
 * @author Yann Hodiesne
 *
 * @param {object} error : The error to handle.
 *
 * @returns {Promise.reject}
 */
export const handleError = (error) => {
	const {
		status,
		statusText: message,
	} = error;

	return Promise.reject({ status, message });
};

/**
 * @function
 * @name post
 * @description Creates a POST method http call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} url	: The url to target.
 * @param {object} body	: The body to include.
 *
 * @returns {Promise | object}
 */
export const post = (url, body) => fetch(`${baseUrl}${url}`, {
	method: 'POST',
	body: JSON.stringify(body),
	headers: getHeaders(),
}).then(handleResponse).catch(handleError);

/**
 * @function
 * @name get
 * @description Creates a GET method http call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} url		: The url to target.
 * @param {objcet} qsObject	: The object to convert into a query string.
 *
 * @returns {Promise | object}
 */
export const get = (url, qsObject) => {
	const queryString = (qsObject && Object.keys(qsObject).length !== 0) ? `?${objectToQS(qsObject)}` : '';

	return fetch(`${baseUrl}${url}${queryString}`, {
		method: 'GET',
		headers: getHeaders(),
	}).then(handleResponse).catch(handleError);
};

/**
 * @function
 * @name put
 * @description Creates a PUT method http call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} url : The url to target.
 * @param {object} body : The body to include.
 */
export const put = (url, body) => fetch(`${baseUrl}${url}`, {
	method: 'PUT',
	body: JSON.stringify(body),
	headers: getHeaders(),
}).then(handleResponse).catch(handleError);

/**
 * @function
 * @name del
 * @description Creates a DELETE method http call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} url : The url to target.
 */
export const del = (url) => fetch(`${baseUrl}${url}`, {
	method: 'DELETE',
	headers: getHeaders(),
}).then(handleResponse).catch(handleError);
