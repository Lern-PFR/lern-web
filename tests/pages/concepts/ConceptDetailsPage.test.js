import { mount, shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import routeData from 'react-router';
import fetchMock from 'fetch-mock';

import RouterProvider from 'routes/components/RouterProvider';
import { ActionTypes as ConceptActionTypes } from 'redux/actions/concepts';

import { ConceptDetailsPage } from 'pages/concepts';
import { baseUrl } from 'lib/shared/http';

describe('Concept details page', () => {
	const mockStore = configureMockStore([thunk]);
	let store;

	beforeEach(() => {
		jest.spyOn(routeData, 'useParams').mockReturnValue({ conceptId: 'dummy_concept_id' });
		store = mockStore({
			concepts: {
				items: [
					{
						id: 'dummy_concept_id',
						title: 'dummy concept title',
						description: 'dummy concept desc',
						moduleId: 'dummy_module_id',
						lessons: [
							{
								id: 'dummy_lesson_id',
								title: 'dummy_lesson_title',
								description: 'dummy_lesson_desc',
								content: 'dummy_lesson_content',
								conceptId: 'dummy_concept_id',
								order: 0,
							},
						],
					},
				],
			},
		});
	});

	afterEach(() => {
		jest.restoreAllMocks();
		jest.resetAllMocks();
	});

	describe('Snapshot testing', () => {
		it('should match previous snapshot without the required concept in state', () => {
			store = mockStore({});

			const sut = (
				<Provider store={store}>
					<RouterProvider>
						<ConceptDetailsPage />
					</RouterProvider>
				</Provider>
			);
			const wrapper = shallow(sut);

			expect(wrapper).toMatchSnapshot();
		});

		it('should match previous snapshot with the required concept in state', () => {
			const sut = (
				<Provider store={store}>
					<RouterProvider>
						<ConceptDetailsPage />
					</RouterProvider>
				</Provider>
			);
			const wrapper = shallow(sut);

			expect(wrapper).toMatchSnapshot();
		});
	});

	describe('dispatched method calls', () => {
		it('should call the fetchConcept action creator on mount', () => {
			const httpResponse = {
				status: 200,
				body: { id: 'dummy_concept_id', title: 'dummy_concept_title', description: 'dummy_concept_description', moduleId: 'dummy_module_id' },
				headers: { 'content-type': 'application/json' },
			};

			fetchMock.get(`${baseUrl}/api/concepts/dummy_concept_id`, httpResponse);

			mount(
				<Provider store={store}>
					<RouterProvider>
						<ConceptDetailsPage />
					</RouterProvider>
				</Provider>
			);

			expect(store.getActions()).toContainEqual({ type: ConceptActionTypes.FETCH_CONCEPT_REQUEST });
		});
	});
});
