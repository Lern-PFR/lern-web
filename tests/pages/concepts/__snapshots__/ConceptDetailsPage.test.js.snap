// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Concept details page Snapshot testing should match previous snapshot with the required concept in state 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <RouterProvider>
    <ConceptDetailsPage />
  </RouterProvider>
</ContextProvider>
`;

exports[`Concept details page Snapshot testing should match previous snapshot without the required concept in state 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <RouterProvider>
    <ConceptDetailsPage />
  </RouterProvider>
</ContextProvider>
`;
